pytorch_lightning/trainer/supporters.py

```diff
@@ -120,26 +120,106 @@ def mean(self):
-    def __init__(self, global_rank: int, world_size: int):
+
+    ID_KEY = 'id'
+
+    def __init__(self, global_rank: int, world_size: int, all_gather_fn: Callable):
         self.global_rank = global_rank
         self.world_size = world_size
-        self.predictions = {}
-        self.num_predictions = 0
+        self.all_gather_fn = all_gather_fn
+        self._legacy_predictions = {}
+        self._predictions = {stage: {} for stage in LoggerStages}
+
+    @property
+    def predictions(self) -> Dict:
+        return self._predictions[self.current_stage]
+
+    def _cache_prediction(self, predictions: List[Dict], dl_idx: int, batch_indices: List[int]) -> None:
+        cache = self.predictions
+
+        cache.setdefault(dl_idx, {})
+
+        def convert(value):
+            return value.cpu().tolist()
+
+        for batch_index, pred in zip(batch_indices, predictions):
+            if batch_index in cache[dl_idx]:
+                raise MisconfigurationException(
+                    "Prediction Collection doesn't support multiple predictions for one sample yet.")
+
+            cache[dl_idx][batch_index] = {
+                self.ID_KEY: batch_index, "predictions": apply_to_collection(pred, torch.Tensor, convert)}
+
+    def cache(self, predictions: List, dl_idx: int, batch_indices: List[int], current_stage: str) -> None:
+        """
+        This function expects predictions to be a list of tensors or dictionnaries of tensors.
+        Example::
+
+            self.add_predictions(predictions)
+        """
+        if predictions is None or (isinstance(predictions, list) and not predictions):
+            return
+
+        self.current_stage = current_stage
+
+        assert isinstance(predictions, (list, torch.Tensor))
+        assert len(predictions) == len(batch_indices)
+        self._cache_prediction(predictions, dl_idx, batch_indices)
+
+    def attach_predictions(self, results: List[Dict], current_stage: str) -> List[Dict]:
+        # indicates the current
+        self.current_stage = current_stage
+        predictions = self.predictions
+        for dl_idx, result in enumerate(results):
+            if dl_idx in predictions:
+                dl_predictions = predictions[dl_idx]
+                dl_predictions = self.all_gather_predictions(dl_predictions)
+                result["predictions"] = list(dl_predictions.values())
+        return results
+
+    def all_gather_predictions(self, predictions: Dict) -> Dict:
+        """
+        This function all_gather predictions accross multiple processes
+        # todo: see https://github.com/PyTorchLightning/pytorch-lightning/issues/5493 for better details.
+        """
+        if not (torch.distributed.is_available() and torch.distributed.is_initialized()):
+            return predictions
+
+        predictions = self.all_gather_fn(predictions)
+
+        def gather(pred: Union[List[Tensor], Tensor], idx: int) -> Union[List[Tensor], Tensor]:
+            def convert(p):
```

Comment:
Some explanation here would be good as to what's happening for future debugging

Reply:
Yes, you are entirely right !

Reply:
Done !